File: Unstop/Practice Problems/Medium/toomanybooks.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int max_books_on_shelf(int n, vector<int> &heights)
{
    vector<int> dp(n + 1, 0); // Initialize dp array

    for (int i = 1; i <= n; ++i)
    {
        dp[i] = 1; // At least one book on the shelf
        for (int j = i - 1; j >= 1; --j)
        {
            if (heights[i - 1] > heights[j - 1])
            {
                dp[i] = max(dp[i], dp[j] + 1);
            }
        }
    }

    return *max_element(dp.begin(), dp.end()); // Return the maximum value in dp array
}

int main()
{
    int n;
    cin >> n;

    vector<int> heights(n);
    for (int i = 0; i < n; ++i)
    {
        cin >> heights[i];
    }

    cout << max_books_on_shelf(n, heights) << endl;

    return 0;
}

File: Unstop/Practice Problems/Medium/toomanybooks.py
def max_books_on_shelf(n, heights):
    dp = [0] * (n + 1)  # Initialize dp array
    dp[0] = 0  # Base case

    for i in range(1, n + 1):
        dp[i] = 1  # At least one book on the shelf
        for j in range(i - 1, 0, -1):
            if heights[i - 1] > heights[j - 1]:
                dp[i] = max(dp[i], dp[j] + 1)

    return max(dp)  # Return the maximum value in dp array

# Input number of books
n = int(input())

# Input heights of books
heights = list(map(int, input().split()))

print(max_books_on_shelf(n, heights))